### DISCUSSION TITLE
What is the correct format for audio samples in the STK "Drummer"?

### DISCUSSION THREAD

**Discussion by KontinuumLab** [pschatzmann/arduino-audio-tools#1196] [Q&A] [answered]
Hi everyone. Great community here, already solved several problems by digging through old threads. Thanks Phil, for being so helpful.

I'm working on an electronic drum kit, based on the ESP32 and using the Audio-tools library, but I need to add several samples with other drum sounds to the STK "Drummer". I'm able to add samples by putting them in the file structure of the library, and then making a few changes to "Drummer.h", "Drummer.cpp" and "all.h". 

The problem is that the library needs samples in "uint16  in 1 channel at 22050 Hz" (information that I found in the library file "MemoryWvin.h") and I haven't found a way to make these files. I'm able to use samples in other formats, but they just sound like terrible screeching and scratching when played.

Here's how I currently make my sample files (in Linux):
1: Open Wav file in Audacity. Mix down to mono and resample to 22500. 
2: Export as RAW (header-less). **This is the problem step, as Audacity doesn't let me export as 16-bit unsigned (only 16-bit signed, or  8-bit signed & unsigned)**
3: Open RAW file in Okteta and export as a C-array. Okteta lets me export the array in whatever format I want, but I still need the original RAW file to have the correct format.

Hope someone has a viable way to do this.
Thanks!

**Comment by pschatzmann** [comment, marked as answer]
I think 16 bits signed is what you need! There is also some information in the [Wiki](https://github.com/pschatzmann/arduino-audio-tools/wiki/MemoryStream:-Converting-a-File-to-Flash-Memory)

**Comment by KontinuumLab** [comment]
Hi Phil. Thanks for getting back to me.
I understand that signed ints are the standard, and I've been able to play my own (signed int) samples through the "streams-memory_raw-i2s" example, using the steps mentioned above.
However, when I tried using those same samples in the Drummer, all I got was a horrible screeching sound. So after digging through the library and trying to figure out the problem, I eventually went to the file: "MemoryWvIn.h" which is included in "Drummer.h", where I found this comment:
/***************************************************/
/*! \class stk::MemoryWvIn
    \brief STK audio input class from Memory

    Class which simulates the FileWvIn class so that it can
    be used by microcontrollrs. We only support data
    recorded as uint16  in 1 channel at 22050 Hz!

    A 'virtual' file can be registered so that a fileOpen
    call with a file name is working like in the corresponding
    FileWvIn class

    This is basically a simplified c of the FileWvIn

    by Phil Schatzmann
*/
/***************************************************/

It seems to specifically be asking for "uint16", which would be unsigned, right? 
I don't pretend to know why this would be the case here, but I think this is where my problem stems from. 
Any ideas?

**Comment by pschatzmann** [comment]
You could write your own small conversion tool which just reads the signed values (-32768 to 32767) and adds 32768 to convert them to unsigned (0 to 65535).
What functionality does the Drummer provide that is missing in MemoryStream ?

**Reply by KontinuumLab** [reply]
The plot thickens. 
Looking closer at the actual built in samples that the Drummer uses, they are actually unsigned chars, while "MemoryWvin" clearly asks for uint16. 
After trying every kind of format of sample I can make (including unsigned chars), I'm still unable to get anything but screeching noises out of the Drummer with anything but the built in samples. Very weird and extremely frustrating. 
I might be missing something obvious.... **Has anyone been able to add a fresh sample to the Drummer??**

The cool thing about the Drummer vs MemoryStream is that it takes care of polyphony automatically and cleanly, which I'm having a very hard time doing in the MemoryStream (using the mixer etc.) 
Also, the Drummer has a nice and simple "NoteOn/NoteOff" interface, which is neat.

**Comment by pschatzmann** [comment]
When you try to import the data (e.g. bassdrum.raw) from the rawwaves directory into Audacity, you will see that it's signed 16 bit, __big endian__, mono!

![image](https://github.com/pschatzmann/arduino-audio-tools/assets/2078916/f71d2b77-8914-4131-a837-3492376a2bb6)

So you need to convert your data from little to big endian!

**Reply by KontinuumLab** [reply]
Yes!
Big endian / little endian was the problem! Thank you Phil.

So for future reference in case anyone comes up against this, here's how you can make samples that work with the STK Drummer:
1: Open Wav file in Audacity. Mix down to mono and resample to 22500.
2: Export audio as RAW (header-less), signed 16-bit PCM.
3: Open RAW file in Okteta or other hex editor and export as a C-array (unsigned char)
4: **Convert the array to big endian.** The 16-bit data is split up into pairs of 8-bit chars, and the conversion consists in flipping the order of each pair. So the first char becomes the second, while the second becomes first, 3rd and 4th become 4th and 3rd etc. I'm doing this directly in an Arduino sketch, but python would probably be a good choice for this operation.

I tried to unmark the other answer so I could mark this one instead, as this is where the mystery was finally solved for me at least, but I wasn't able.
Thanks again Phil!
